Speed up dockerized builds

jbeda opened the pull request:
This PR speeds up dockerized builds.  First, we make sure that we are as incremental as possible.  The bigger change is that now we use rsync to move sources into the container and get data back out.

To do yet:
- [x] Add a random password to rsync.  This is 128bit MD4, but it is better than nothing.
- [x] Lock down rsync to only come from the host.
- [x] Deal with remote docker engines -- this should be necessary for docker-machine on the mac.
- [x] Allow users to specify the port for the rsync daemon.  Perhaps randomize this or let docker pick an ephemeral port and detect the port?
- [x] Copy back generated files so that users can check them in.  This is done for `zz_generated.*` files generated by `make generated_files` 
  - [x] This should include generated proto files so that we can remove the hack-o-rama that is `hack/hack/update-*-dockerized.sh` 
- [x] Start "versioning" the build container and the data container so that the CI system doesn't have to be manually kicked.
- [x] Get some benchmarks to qualify how much faster.

This replaces #28518 and is related to #30600.

cc @thockin @spxtr @david-mcmahon @MHBauer 

Benchmarks by running `make clean ; sync ; time bash -xc 'time build/make-build-image.sh ; time sync ; time build/run.sh make ; time sync; time build/run.sh make'` on a GCE n1-standard-8 with PD-SSD.

| setup | build image | sync | first build | sync | second build | total |
| --- | --- | --- | --- | --- | --- | --- |
| baseline | 0m11.420s | 0m0.812s | 7m2.353s | 0m42.380s | 7m8.381s | 15m5.348s |
| this pr | 0m10.977s | 0m15.168s | 7m31.096s | 1m55.692s | 0m16.514s | 10m9.449s |

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30787)

<!-- Reviewable:end -->

jbeda committed:
Remove support for boot2docker

jbeda committed:
Make update-bindata be incremental build friendly

jbeda committed:
Use rsync to get source into build container

We also add "version" to all docker images and containers

This version is to be incremented manually when we change the shape of the build
image (like changing the golang version or the set of volumes in the data
container).  This will delete all older versions of images and containers when
the version is different.

spxtr commented:
Roughly how much faster is it?

jbeda commented:
I'll pull together some benchmarks before I remove the WIP label.

thockin commented on build/common.sh:
```diff
@@ -566,14 +533,38 @@ function kube::build::ensure_data_container() {
-      chown -R $(id -u).$(id -g) "${REMOTE_OUTPUT_ROOT}"
+      chown -R $(id -u).$(id -g)
+        "${REMOTE_ROOT}"
+        /usr/local/go/pkg/linux_386_cgo
```
why not just do all of `/usr/local/go/pkg/`

jbeda replied:
That ends up whiting out the compiler:
```
$ make WHAT='cmd/kubectl'
+++ [0820 16:28:15] Generating bindata:
    /go/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
../../../hack/../test/e2e/generated/bindata.go.tmp ../../../hack/../test/e2e/generated/bindata.go differ: char 1523, line 29
+++ [0820 16:28:16] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
go tool: no such tool "compile"
Makefile.generated_files:279: recipe for target '_output/bin/deepcopy-gen' failed
make[1]: *** [_output/bin/deepcopy-gen] Error 1
Makefile:281: recipe for target 'generated_files' failed
make: *** [generated_files] Error 2
```

thockin replied:
to just chown?  Or is that from volume mounting on pkg, which I could understand.

jbeda replied:
The compiler itself is under /usr/local/go/pkg.  If we make that a volume it gets whited out.  Not sure what else to do there.

We could do something where we move it aside when building the image and then copy it back as part of the chown step, but that seems complicated too.

thockin replied:
I meant just to chmod/chown the whole thing, not to volume mount the whole
thing.

On Mon, Aug 22, 2016 at 2:24 PM, Joe Beda <notifications@github.com> wrote:

> In build/common.sh
> <https://github.com/kubernetes/kubernetes/pull/30787#discussion_r75761703>
> :
>
> >        --volume "${REMOTE_OUTPUT_GOPATH}" # make a non-root owned mountpoint
> >        "${KUBE_BUILD_IMAGE}"
> > -      chown -R $(id -u).$(id -g) "${REMOTE_OUTPUT_ROOT}"
> > +      chown -R $(id -u).$(id -g)
> > +        "${REMOTE_ROOT}"
> > +        /usr/local/go/pkg/linux_386_cgo
>
> The compiler itself is under /usr/local/go/pkg. If we make that a volume
> it gets whited out. Not sure what else to do there.
>
> We could do something where we move it aside when building the image and
> then copy it back as part of the chown step, but that seems complicated too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30787/files/c0ba418d1f5f07f74ae8e9aa392e7ddbcfa0a144#r75761703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEMQP4zlhDmbqAsqoZD3z7Rl7Y3Yks5qihN5gaJpZM4Jmt7e>
> .
>

jbeda replied:
That seems to work.  Latest versions have that.

thockin commented on build/common.sh:
```diff
@@ -582,19 +573,49 @@ function kube::build::ensure_data_container() {
+#
+# Arguments are in the form of
+#  <container name> <extra docker args> -- <command>
+function kube::build::run_build_command_ex() {
```
"ex"?  Your microsoft is showing, Joe

jbeda replied:
Open to other naming conventions here :)

thockin replied:
just ribbing :)

jbeda committed:
Move release code (bash) out of build/common.sh

jbeda committed:
Make update-generated-*.sh not hack around with docker images

make-generated-{protobuf,runtime}.sh was doing some really nasty stuff with how
the build container was managed in order to copy results out.  Since we have
more flexibility to grab results out of the build container, we can now avoid
all of this.  Ideally we wouldn't have `hack` calling `build` at all, but we
aren't there yet.

thockin commented on cmd/libs/go2idl/conversion-gen/generators/conversion.go:
```diff
@@ -280,8 +280,7 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
 		if !pkgNeedsGeneration {
-			glog.V(5).Infof("  no viable conversions, not generating for this package")
-			continue
+			glog.Errorf("Package %v has 'k8s:conversion-gen' tags but no viable conversions", i)
```
seems unrelated...

jbeda replied:
The fact that this wasn't outputting a file was confusing your generated files make system.  I wasn't able to get to a "no op" make without some sort of fix here.

thockin replied:
??  How do I repro that at head?

On Mon, Aug 22, 2016 at 6:18 AM, Joe Beda <notifications@github.com> wrote:

> In cmd/libs/go2idl/conversion-gen/generators/conversion.go
> <https://github.com/kubernetes/kubernetes/pull/30787#discussion_r75674739>
> :
>
> > @@ -280,8 +280,7 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
> >  			}
> >  		}
> >  		if !pkgNeedsGeneration {
> > -			glog.V(5).Infof("  no viable conversions, not generating for this package")
> > -			continue
> > +			glog.Errorf("Package %v has 'k8s:conversion-gen' tags but no viable conversions", i)
>
> The fact that this wasn't outputting a file was confusing your generated
> files make system. I wasn't able to get to a "no op" make without some sort
> of fix here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30787/files/9269652b4fc9760754e0b4e47cc2cb995f9d3e24#r75674739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVM2XusCz26zLyZDn5qPbaCQ91KPbks5qiaGdgaJpZM4Jmt7e>
> .
>

jbeda replied:
It's pretty fast but still extra work.

<details><summary>repo</summary>
```
[jbeda@k8s-build:~/src/kubernetes] (master)$ make clean
+++ [0822 18:08:36] Verifying Prerequisites....
!!! [0822 18:08:36] Build image not built.  Cannot clean via docker build image.
+++ [0822 18:08:36] Removing data container kube-build-data-27932f596f
+++ [0822 18:08:36] Removing _output directory
+++ [0822 18:08:36] Deleting docker image kube-build:build-27932f596f
+++ [0822 18:08:36] Cleaning all other untagged docker images
:) Mon Aug 22 18:08:36
[jbeda@k8s-build:~/src/kubernetes] (master)$ make generated_files
+++ [0822 18:08:49] Generating bindata:
    /home/jbeda/src/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0822 18:08:49] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0822 18:08:50] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [0822 18:08:55] Generating bindata:
    /home/jbeda/src/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0822 18:08:56] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0822 18:08:56] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
E0822 18:09:00.238473    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/apps/v1alpha1.PetSetSpec <-> k8s.io/kubernetes/pkg/apis/apps.PetSetSpec
E0822 18:09:00.258650    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.HorizontalPodAutoscalerSpec <-> k8s.io/kubernetes/pkg/apis/autoscaling.HorizontalPodAutoscalerSpec
E0822 18:09:00.266875    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.JobSpec <-> k8s.io/kubernetes/pkg/apis/batch.JobSpec
E0822 18:09:00.282497    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.RollingUpdateDeployment <-> k8s.io/kubernetes/pkg/apis/extensions.RollingUpdateDeployment
E0822 18:09:00.285592    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.ScaleStatus <-> k8s.io/kubernetes/pkg/apis/extensions.ScaleStatus
:) Mon Aug 22 18:09:00
[jbeda@k8s-build:~/src/kubernetes] (master)$ make generated_files --debug=b
GNU Make 4.0
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating goal targets....
 File 'generated_files' does not exist.
Must remake target 'generated_files'.
GNU Make 4.0
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Reading makefiles...
Updating goal targets....
 File 'generated_files' does not exist.
   File 'gen_deepcopy' does not exist.
  Must remake target 'gen_deepcopy'.
  Successfully remade target file 'gen_deepcopy'.
   File 'gen_conversion' does not exist.
     File 'federation/apis/core/v1/zz_generated.conversion.go' does not exist.
    Must remake target 'federation/apis/core/v1/zz_generated.conversion.go'.
    Successfully remade target file 'federation/apis/core/v1/zz_generated.conversion.go'.
  Must remake target 'gen_conversion'.
  Successfully remade target file 'gen_conversion'.
Must remake target 'generated_files'.
Successfully remade target file 'generated_files'.
Successfully remade target file 'generated_files'.
```
</details

thockin replied:
ACK.  Good catch.

On Mon, Aug 22, 2016 at 11:13 AM, Joe Beda <notifications@github.com> wrote:

> In cmd/libs/go2idl/conversion-gen/generators/conversion.go
> <https://github.com/kubernetes/kubernetes/pull/30787#discussion_r75728857>
> :
>
> > @@ -280,8 +280,7 @@ func Packages(context *generator.Context, arguments *args.GeneratorArgs) generat
> >  			}
> >  		}
> >  		if !pkgNeedsGeneration {
> > -			glog.V(5).Infof("  no viable conversions, not generating for this package")
> > -			continue
> > +			glog.Errorf("Package %v has 'k8s:conversion-gen' tags but no viable conversions", i)
>
> It's pretty fast but still extra work.
>
> repo
>
> [jbeda@k8s-build:~/src/kubernetes] (master)$ make clean
> +++ [0822 18:08:36] Verifying Prerequisites....
> !!! [0822 18:08:36] Build image not built.  Cannot clean via docker build image.
> +++ [0822 18:08:36] Removing data container kube-build-data-27932f596f
> +++ [0822 18:08:36] Removing _output directory
> +++ [0822 18:08:36] Deleting docker image kube-build:build-27932f596f
> +++ [0822 18:08:36] Cleaning all other untagged docker images
> :) Mon Aug 22 18:08:36
> [jbeda@k8s-build:~/src/kubernetes] (master)$ make generated_files
> +++ [0822 18:08:49] Generating bindata:
>     /home/jbeda/src/kubernetes/test/e2e/framework/gobindata_util.go
> +++ [0822 18:08:49] Building the toolchain targets:
>     k8s.io/kubernetes/hack/cmd/teststale
> +++ [0822 18:08:50] Building go targets for linux/amd64:
>     cmd/libs/go2idl/deepcopy-gen
> +++ [0822 18:08:55] Generating bindata:
>     /home/jbeda/src/kubernetes/test/e2e/framework/gobindata_util.go
> +++ [0822 18:08:56] Building the toolchain targets:
>     k8s.io/kubernetes/hack/cmd/teststale
> +++ [0822 18:08:56] Building go targets for linux/amd64:
>     cmd/libs/go2idl/conversion-gen
> E0822 18:09:00.238473    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/apps/v1alpha1.PetSetSpec <-> k8s.io/kubernetes/pkg/apis/apps.PetSetSpec
> E0822 <http://k8s.io/kubernetes/pkg/apis/apps.PetSetSpecE0822> 18:09:00.258650    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.HorizontalPodAutoscalerSpec <-> k8s.io/kubernetes/pkg/apis/autoscaling.HorizontalPodAutoscalerSpec
> E0822 <http://k8s.io/kubernetes/pkg/apis/autoscaling.HorizontalPodAutoscalerSpecE0822> 18:09:00.266875    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.JobSpec <-> k8s.io/kubernetes/pkg/apis/batch.JobSpec
> E0822 <http://k8s.io/kubernetes/pkg/apis/batch.JobSpecE0822> 18:09:00.282497    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.RollingUpdateDeployment <-> k8s.io/kubernetes/pkg/apis/extensions.RollingUpdateDeployment
> E0822 <http://k8s.io/kubernetes/pkg/apis/extensions.RollingUpdateDeploymentE0822> 18:09:00.285592    8625 conversion.go:589] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.ScaleStatus <-> k8s.io/kubernetes/pkg/apis/extensions.ScaleStatus
> :) Mon Aug 22 18:09:00
> [jbeda@k8s-build:~/src/kubernetes] (master)$ make generated_files --debug=b
> GNU Make 4.0
> Built for x86_64-pc-linux-gnu
> Copyright (C) 1988-2013 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> Reading makefiles...
> Updating goal targets....
>  File 'generated_files' does not exist.
> Must remake target 'generated_files'.
> GNU Make 4.0
> Built for x86_64-pc-linux-gnu
> Copyright (C) 1988-2013 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> Reading makefiles...
> Updating goal targets....
>  File 'generated_files' does not exist.
>    File 'gen_deepcopy' does not exist.
>   Must remake target 'gen_deepcopy'.
>   Successfully remade target file 'gen_deepcopy'.
>    File 'gen_conversion' does not exist.
>      File 'federation/apis/core/v1/zz_generated.conversion.go' does not exist.
>     Must remake target 'federation/apis/core/v1/zz_generated.conversion.go'.
>     Successfully remade target file 'federation/apis/core/v1/zz_generated.conversion.go'.
>   Must remake target 'gen_conversion'.
>   Successfully remade target file 'gen_conversion'.
> Must remake target 'generated_files'.
> Successfully remade target file 'generated_files'.
> Successfully remade target file 'generated_files'.
>
> </details
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30787/files/9269652b4fc9760754e0b4e47cc2cb995f9d3e24#r75728857>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVNEJWxeUo7Q5hXJa-bOF4I1RVespks5qiebagaJpZM4Jmt7e>
> .
>

jbeda commented:
To anyone watching from home, I'm fighting with Jenkins.  I was using a Docker 1.12 feature and had to discover that Jenkins was using 1.9.1 and adjust.  Now the verification test is failing (can't talk to the rsync container) and I'm trying to figure out why.  The only way I have to test is to push a new change with debug commands.  It is essentially printf debugging with a 15m+ delay.

MHBauer commented on build/README.md:
```diff
@@ -4,21 +4,18 @@ Building Kubernetes is easy if you take advantage of the containerized build env
-3. **Optional** [Google Cloud SDK](https://developers.google.com/cloud/sdk/)
+1. Docker, using one of the following configurations:
+  1. **Mac OS X** You can either use Docker for Mac or docker-machine. See installation instructions [here](https://docs.docker.com/installation/mac/).
+     **Note**: You will want to set the Docker VM to have at least 3GB of initial memory or building will likely fail. (See: [#11852]( http://issue.k8s.io/11852)). https://github.com/kubernetes/kubernetes/issues/14773))
```
Too many closing parens on this line.

jbeda commented:
It looks like something is messed up on the Docker install on the Jenkins machine running verify.  When running `ip addr` inside a container we get this:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
804: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc noqueue state DOWN group default 
    link/ether 02:42:ac:11:00:08 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.8/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:8/64 scope link tentative 
       valid_lft forever preferred_lft forever
```

It looks like the link inside the container is down for some reason.  Without looking at other logs, it is difficult to figure out why this is the case.  I'm going to need someone with access to the machine to move this forward.

ixdy commented:
I don't think it's that machine in particular.
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/30787/kubernetes-pull-verify-all/10789/build-log.txt and https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/30787/kubernetes-pull-verify-all/11059/build-log.txt were run on different Jenkins agents, but the failures look similar.

The verify tests are using docker-in-docker, which is known to be mysterious. I'm guessing that's probably what's wrong.

Maybe try emulating Jenkins locally?
```
WORKSPACE=/tmp/dummy KUBE_FORCE_VERIFY_CHECKS='y' KUBE_TEST_SCRIPT="./hack/jenkins/verify-dockerized.sh" ./hack/jenkins/gotest-dockerized.sh
```

MHBauer commented on build/common.sh:
```diff
@@ -629,977 +581,117 @@ function kube::build::run_build_command() {
+    --filter='+ *.pb.go' \
+    --filter='+ */' \
+    --filter='- /**' \
+    --archive --perms \
```
doesn't `archive` imply `perms`?

jbeda replied:
fixed

MHBauer commented:
Working on OSX.

On a linux vm, I am having the incredibly bizarre problem where I cannot rsync things into the container. rsync complains about not having enough permissions. Doesn't make any sense as rsync is running as root in the container.

jbeda commented:
@MHBauer -- very strange.  Which version of Docker are you using? Are you running it as root or as your normal user?  I had it working but I'll re-verify.  Which rsync is complaining? Is it the daemon (as seen by `docker logs`) or is it the client (as seen by the output of the script)? Is it happening when syncing into the container or copying stuff out?

MHBauer commented:
daemon side, rsyncing into the container. Running as root. Docker v1.12.1 on both OSX and Linux. I'll try as a user, and another version of docker as well.

errors look like
```
rsync: failed to set times on "." (in k8s): Operation not permitted (1)
rsync: recv_generator: mkdir ".github" (in k8s) failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
rsync: recv_generator: mkdir "Godeps" (in k8s) failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
```
I **really** want this to work. I was very surprised my Linux VM had a problem.

MHBauer commented:
It was "Don't run as root". As a normal user it works fine.

jbeda commented:
@MHBauer -- thanks for the report.  I'll make it work as root before I check in.

MHBauer commented:
Writing down my understanding of the setup:

- kube-build-data container that holds all the build data in volumes. 
- rsync-daemon container that mounts the volumes from the build-data container, for copying things in-to and out-of the build-data volumes
- make-build container that runs make

resulting in sequence:

1. create data container
2. create and start rsync container
3. copy local source into build container through the rsync container
4. rsync goes away
5. start build container running make, run to completion
6. rsync start to copy results out
7. rsync goes away
8. build container and volumes are left around until next build.

If so, looks cool to me. Gives me some other ideas to look at in the future.

jbeda commented:
Mostly right -- the build container is also deleted and restarted as necessary. The only container that stays around is the data container.  As it is created a simple "chown" is run on some dirs but then it stays around dead most of the time.

jbeda commented:
ok to test

jbeda commented:
@ixdy -- this should be ready to do a full review on now.  It should be passing all of the Jenkins tests.

MHBauer commented:
@jbeda flawless runs for me. root & non-root. osx & linux.

os | build | rebuild
--- | ---- | ---
osx | 685s | 65s
linux | 1121s | 55s

Will try to review. Do you prefer github comments or the purple zone?

jbeda commented:
@MHBauer Whatever is cool with me. Glad to see rebuild is reasonable.  Is your linux machine significantly slower than your mac?

MHBauer commented:
Ya, 1core vm down in the basement somewhere vs 4core mac. Sorry if that makes it look weird. We can see `make`'s dependency management or whatever working. Ripping fast disk mount. May be time to ask for new VM.

MHBauer commented on build/common.sh:
```diff
@@ -274,15 +249,15 @@ function kube::build::ensure_docker_daemon_connectivity {
+      echo "        - docker-machine create -d ${DOCKER_MACHINE_DRIVER} ${DOCKER_MACHINE_NAME}"
+      echo "      - Set your environment variables using: "
+      echo "        - eval \$(docker-machine env ${DOCKER_MACHINE_NAME})"
+      echo "      - Update your Docker VM"
```
I don't agree with this indentation change.
While the `upgrade` text is docker-machine specific, the 'update your docker vm' case is not a sub section of the 'set up docker-machine' bullet.

Alternatively, clarify the text with something like "If you see the Error Message ....  your docker is out of date."

I'm not sure how much discussion helptext we should have here. Possibly it should be a link to some 'basic development environment troubleshooting'.

jbeda replied:
Rewrote it a bit. Take another look after I push.  I like the idea of a trouble shooting page, but it is so easy to get disconnected.

MHBauer commented on hack/update-bindata.sh:
```diff
@@ -37,14 +36,21 @@ if ! which go-bindata &>/dev/null ; then
+gofmt -s -w "${BINDATA_OUTPUT}.tmp"
 
-V=2 kube::log::info "Generated bindata file : $(wc -l ${BINDATA_OUTPUT}) lines of lovely automated artifacts"
+if ! cmp  "${BINDATA_OUTPUT}.tmp" "${BINDATA_OUTPUT}" ; then
```
compare and overwrite only if different to avoid always overwriting. 

Because it changes the file timestamps?

Why use `cat` and not `mv` or `cp`?

jbeda replied:
Exactly -- this was triggering a rebuild more often than needed.  

`cp` will maintain file permissions of the source file while `cat >` will maintain file permissions of the target file (it does a truncate and then a write).  I'm following the pattern that @thockin is using in `Makefile.generated_files`

MHBauer replied:
Okay cool. Glad I understand. I'm not a bash wizard, so it stood out to me. Could we get a tiny little comment with the basic explanation?

```
compare and overwrite only if different to avoid updating the timestamp and triggering make. cat redirect trick to preserve file permissions of the target file.
```

Thanks. I am done with this comment whether or not it is changed.

jbeda replied:
Done in latest push.

MHBauer commented on build/build-image/rsyncd.sh:
```diff
@@ -0,0 +1,80 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file will set up and run rsyncd in order to sync kubernetes sources back
```
can you add a little more info like "in order to sync kubernetes source between the local filesystem and the docker-build-data container"

I see this sentence and I ask myself, 'sync source back and forth between what and what?'

jbeda replied:
Added some more detail.

MHBauer commented on build/common.sh:
```diff
@@ -55,7 +55,11 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
+# KUBE_DATA_CONTAINER_NAME=kube-build-data-<hash>"
+
+# Increment this version to cause everyone to rebuild images, including the data
+# container.
```
How often should we be doing this version change? My guess is major versions. Is this 5 for k8s v1.5?

jbeda replied:
This is necessary when we do major changes to the build environment.  Things like bump the golang version or add new volumes to the data container.  We now also no longer `docker build` on every `build/run.sh` so if you need to rebuild the build container (say, you modify `rsyncd.sh`) then you'd want to increment this too.  This doesn't line up with major kubernetes versions.

I'll add some more text explaining this.

MHBauer commented on build/common.sh:
```diff
@@ -596,70 +653,134 @@ function kube::build::run_build_command() {
-  kube::build::run_build_command touch "${REMOTE_OUTPUT_BINPATH}/test_for_remote"
+function kube::build::probe_address {
+  # Apple has an ancient version of netcat with custom timeout flags.  This is
+  # the best way I (jbeda) could find to test for that.
```
No actionable comment here.

I am also disappointed in the old tools. If we can't force people to `brew install`, then another reason for moving into a container is to get rid of any kind of platform specific scripting.

thockin replied:
+1 - I see MacOS as MORE impetus to make the docker-build be default.

On Wed, Aug 31, 2016 at 6:20 PM, Morgan Bauer <notifications@github.com>
wrote:

> In build/common.sh
> <https://github.com/kubernetes/kubernetes/pull/30787#discussion_r77101203>
> :
>
> >    kube::build::destroy_container "${KUBE_BUILD_CONTAINER_NAME}"
> >  }
> >
> > -# Test if the output directory is remote (and can only be accessed through
> > -# docker) or if it is "local" and we can access the output without going through
> > -# docker.
> > -function kube::build::is_output_remote() {
> > -  rm -f "${LOCAL_OUTPUT_SUBPATH}/test_for_remote"
> > -  kube::build::run_build_command touch "${REMOTE_OUTPUT_BINPATH}/test_for_remote"
> > +function kube::build::probe_address {
> > +  # Apple has an ancient version of netcat with custom timeout flags.  This is
> > +  # the best way I (jbeda) could find to test for that.
>
> No actionable comment here.
>
> I am also disappointed in the old tools. If we can't force people to brew
> install, then another reason for moving into a container is to get rid of
> any kind of platform specific scripting.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30787/files/487945995c84c43caa6f82e902bf277adf89dad2..166d23a43aa2b0f825bc6feec938684b1f43a323#r77101203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVKFpENUCcO-O8a5FYFVH0ADCcJlkks5qlih7gaJpZM4Jmt7e>
> .
>

jbeda replied:
Totally agree that we should minimize this.  The stuff in `build/` is needed to enable us to easily get there.

My guide: `build/*` should work without having to install anything special besides Docker for Mac.  That is super easy and we can almost assume it at this poing.

Stuff in `hack/*` should feel free to make linux-y/GNU assumptions.  If users *really* want to run on OS X it'll be unsupported and they'll probably have to `brew install` some stuff.  Most users will be told to just do a `build/run.sh <command>` instead.

jbeda committed:
Install test subpackages too

We weren't getting incremental builds because of new test only subpackages.  Our
voodoo combo of 'go install' and 'go test -c' didn't cache things like
'test/e2e_node/services'.  Add the '-i' flag to 'go test' to install test only
dependencies too.

jbeda commented:
Review status: 0 of 26 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[build/common.sh, line 559 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787#-KPOdUJWv0I16HhLW2-9-r1-559:-KQbnvCe7eSzcWRdGu7R:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c0ba418d1f5f07f74ae8e9aa392e7ddbcfa0a144/build/common.sh#L559)):*
<details><summary><i>Previously, jbeda (Joe Beda) wrote…</i></summary>
> That seems to work.  Latest versions have that.
</details>
Done.

---

*[build/common.sh, line 693 \[r19\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787#-KPOdUJWv0I16HhLW2-9-r19-693:-KQbnUDL5S964eGaRJGN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/5057c0ff12e21acf7da93772725af7b1ebfeab47/build/common.sh#L693)):*
<details><summary><i>Previously, jbeda (Joe Beda) wrote…</i></summary>
> fixed
</details>
Done.

---

*[build/common.sh, line 258 \[r20\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787#-KPOdUJWv0I16HhLW2-9-r20-258:-KQbnUs0AwDVQ7k1TZBR:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/09e181967da42bc4810afe150150825d7b27714a/build/common.sh#L258)):*
<details><summary><i>Previously, jbeda (Joe Beda) wrote…</i></summary>
> Rewrote it a bit. Take another look after I push.  I like the idea of a trouble shooting page, but it is so easy to get disconnected.
</details>
Done.

---

*[build/common.sh, line 61 \[r23\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787#-KPOdUJWv0I16HhLW2-9-r23-61:-KQbnVPwHKw5F98618uh:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/166d23a43aa2b0f825bc6feec938684b1f43a323/build/common.sh#L61)):*
<details><summary><i>Previously, jbeda (Joe Beda) wrote…</i></summary>
> This is necessary when we do major changes to the build environment.  Things like bump the golang version or add new volumes to the data container.  We now also no longer `docker build` on every `build/run.sh` so if you need to rebuild the build container (say, you modify `rsyncd.sh`) then you'd want to increment this too.  This doesn't line up with major kubernetes versions.
> 
> 
> 
> I'll add some more text explaining this.
</details>
Done.

---

*[build/common.sh, line 662 \[r23\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787#-KPOdUJWv0I16HhLW2-9-r23-662:-KQbnWMTdLxZZJRX_j9W:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/166d23a43aa2b0f825bc6feec938684b1f43a323/build/common.sh#L662)):*
<details><summary><i>Previously, jbeda (Joe Beda) wrote…</i></summary>
> Totally agree that we should minimize this.  The stuff in `build/` is needed to enable us to easily get there.
> 
> 
> 
> My guide: `build/*` should work without having to install anything special besides Docker for Mac.  That is super easy and we can almost assume it at this poing.
> 
> 
> 
> Stuff in `hack/*` should feel free to make linux-y/GNU assumptions.  If users *really* want to run on OS X it'll be unsupported and they'll probably have to `brew install` some stuff.  Most users will be told to just do a `build/run.sh <command>` instead.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787)*
<!-- Sent from Reviewable.io -->

jbeda commented:
@k8s-bot e2e test this issue #26760

jbeda commented:
@MHBauer I need you to resolve discussions reviewable.

jbeda commented:
@ixdy Let me know and I'll squash the fixup commits.

MHBauer commented:
Reviewed 5 of 13 files at r1, 19 of 19 files at r20.
Review status: 5 of 26 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787)*
<!-- Sent from Reviewable.io -->

MHBauer commented:
Reviewed 1 of 1 files at r21, 2 of 2 files at r22, 9 of 9 files at r23, 8 of 8 files at r24, 4 of 4 files at r25, 2 of 2 files at r26, 1 of 1 files at r27, 1 of 1 files at r28.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30787)*
<!-- Sent from Reviewable.io -->

errordeveloper commented:
Are we pretty close to getting this into master? Would love to try using it!

ixdy commented:
I'm planning to look over it later today.

It's large enough that I don't think we want to merge it until the 1.4 code slush thaws. (Probably want to prioritize it to avoid endless rebasing, though.)

luxas commented:
Maybe mark it `pre-1.5-general` in order to have it prioritized when the gates for v1.5 are opening...

jbeda commented:
Thanks.  In theory this shouldn't change the product but it'd suck if we have to delay a release because the build/release process is broken by this.

ixdy commented:
I'm still trying to review this. I wish you had not put so many changes into one PR.

ixdy commented on build/common.sh:
```diff
@@ -176,14 +176,12 @@ function kube::build::docker_available_on_osx() {
-    if [[ -z "$(which docker-machine)" && -z "$(which boot2docker)" ]]; then
-      kube::log::status "It looks like you're running Mac OS X, yet none of Docker for Mac, docker-machine or boot2docker are on the path."
+    if [[ -z "$(which docker-machine)" ]]; then
+      kube::log::status "It looks like you're running Mac OS X, yet neither Docker for Mac or docker-machine can be found."
```
nit: neither/nor

ixdy commented on hack/update-bindata.sh:
```diff
@@ -37,14 +36,24 @@ if ! which go-bindata &>/dev/null ; then
+# Here we compare and overwrite only if different to avoid updating the
+# timestamp and triggering a rebuild. The 'cat' redirect trick to preserve file
+# permissions of the target file.
+if ! cmp  "${BINDATA_OUTPUT}.tmp" "${BINDATA_OUTPUT}" ; then
```
maybe `cmp -s`?

jbeda replied:
Done.

ixdy commented on hack/update-bindata.sh:
```diff
@@ -37,14 +36,24 @@ if ! which go-bindata &>/dev/null ; then
+gofmt -s -w "${BINDATA_OUTPUT}.tmp"
 
-V=2 kube::log::info "Generated bindata file : $(wc -l ${BINDATA_OUTPUT}) lines of lovely automated artifacts"
+# Here we compare and overwrite only if different to avoid updating the
```
the real fix for this is to only run this when we actually need to. right now I think we're running this far more than we need to.

jbeda replied:
Yeah -- but that is probably more invasive.   This'll help contain the damage.

ixdy commented:
while 8279f9a34f3269b912fbc9cbc44178fbc3af367b `Move release code (bash) out of build/common.sh is great`, it seems wholly unrelated to the goal of this PR - why was it included?

ixdy commented on hack/update-generated-protobuf.sh:
```diff
@@ -18,39 +18,12 @@ set -o errexit
-kube::build::update_dockerfile
-kube::build::docker_build "${KUBE_BUILD_IMAGE}" "${LOCAL_OUTPUT_BUILD_CONTEXT}" 'false'
-kube::build::run_build_command hack/update-generated-protobuf-dockerized.sh "$@"
+"${KUBE_ROOT}/build/run.sh" hack/update-generated-protobuf-dockerized.sh "$@"
```
👍 this is so much better.

ixdy commented:
#24629 recently merged, which nuked most of the release stuff anyway. so if you rebase, this should be a smaller PR now.

jbeda commented:
@k8s-bot test this issue: #32574

jbeda commented:
Sorry about having so much in one PR.  I lot of the smaller changes (like pulling out the release stuff) was incidental and hard to juggle.  I'll try to keep these things smaller in the future :)

ixdy commented on build/README.md:
```diff
@@ -29,8 +29,8 @@ The following scripts are found in the `build/` directory. Note that all scripts
   *  `build/run.sh make test-cmd`: Run CLI tests
-* `build/copy-output.sh`: This will copy the contents of `_output/dockerized/bin` from any remote Docker container to the local `_output/dockerized/bin`.  Right now this is only necessary on Mac OS X with `boot2docker` when your git repo isn't under `/Users`.
-* `build/make-clean.sh`: Clean out the contents of `_output/dockerized` and remove any local built container images.
+* `build/copy-output.sh`: This will copy the contents of `_output/dockerized/bin` from any the Docker container to the local `_output/dockerized/bin`.
```
remove `any`?

ixdy commented on build/README.md:
```diff
@@ -29,10 +27,64 @@ The following scripts are found in the `build/` directory. Note that all scripts
-* `build/shell.sh`: Drop into a `bash` shell in a build container with a snapshot of the current repo code.
-* `build/release.sh`: Build everything, test it, and (optionally) upload the results to a GCS bucket.
+* `build/copy-output.sh`: This will copy the contents of `_output/dockerized/bin` from the Docker container to the local `_output/dockerized/bin`.  This is run automatically as part of `build/run.sh`.
+* `build/make-clean.sh`: Clean out the contents of `_output`, remove any locally built container images and remote the data container.
```
swap "remote" and "the"?

jbeda replied:
s/remote/remove/

ixdy commented on build/common.sh:
```diff
@@ -55,7 +55,16 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
+# Increment/change this number if you change the build image (anything
+# under build/build-image, golang version) or change the set of volumes in the
+# data container.
+readonly KUBE_IMAGE_VERSION=5
```
why start at 5? and why not include BUILD in the name?

ixdy commented on build/common.sh:
```diff
@@ -156,12 +166,22 @@ function kube::build::verify_prereqs() {
   fi
   kube::build::ensure_docker_daemon_connectivity || return 1
 
+  if (( $KUBE_VERBOSE > 6 )); then
```
nit: `${KUBE_VERBOSE}`

ixdy commented on build/common.sh:
```diff
@@ -353,9 +320,51 @@ function kube::build::docker_image_exists() {
+  # In Docker 1.12, we can replace this with
+  #    docker images "$1" --format "{{.Tag}}"
+  for tag in $(docker images kube-build | tail -n +2 | awk '{print $2}') ; do
+    if [[ "$tag" != "$2"* ]] ; then
```
nit: `${tag}` throughout

ixdy commented on build/common.sh:
```diff
@@ -353,9 +320,51 @@ function kube::build::docker_image_exists() {
+  # In Docker 1.12 we can replace this line with
+  #   docker ps -a --format="{{.Names}}"
+  for container in $(docker ps -a | tail -n +2 | awk '{print $NF}') ; do
+    if [[ "$container" != "$1"* ]] ; then
```
nit: `${container}` throughout

ixdy commented on build/common.sh:
```diff
@@ -477,49 +506,48 @@ EOF
+    #
+    # The data container creates volumes for all of the directories that store
+    # intermediates for the Go build. This enables incremental builds across
+    # Docker sessions. The *_cgo paths are re-compiled versions of the go std
```
I'm confused - aren't these part of the crossbuild image?

ixdy replied:
wonder if there's some way we can explode this list based on `KUBE_CLIENT_PLATFORMS` or `KUBE_CROSSPLATFORMS` or whatever. (maybe in a follow-up?)

also, we didn't do this previously - why is it necessary now? just to improve performance?

and when are these directories created?

jbeda replied:
A key problem here is the way we do `KUBE_FASTBUILD` in `hack/golang.sh`.  There we modify the list of cross targets.  That means that we won't create data container correctly if we come in with `KUBE_FASTBUILD` set.  The "right" way to do this is to have a variable for all cross targets we might ever support and then a variable for the cross targets for this build.  That is a bigger change for later.

ixdy commented on build/common.sh:
```diff
@@ -528,19 +556,51 @@ function kube::build::ensure_data_container() {
+# Arguments are in the form of
+#  <container name> <extra docker args> -- <command>
+function kube::build::run_build_command_ex() {
+  [[ $# != 0 ]] || { echo "Invalid input - please specify a the container name." >&2; return 4; }
```
"a the"?

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+function kube::build::probe_address {
+  # Apple has an ancient version of netcat with custom timeout flags.  This is
+  # the best way I (jbeda) could find to test for that.
+  local nc
```
I don't understand this line

jbeda replied:
That is supposed to be `local netcat`.  I'll fix it up.

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+
+  local tries=10
+  while (( $tries > 0 )) ; do
+    if $netcat -z "$1" "$2" 2> /dev/null ; then
```
nit: `${netcat}`

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
 
   # Clean up container from any previous run
   kube::build::destroy_container "${KUBE_BUILD_CONTAINER_NAME}"
```
these are probably supposed to be `${container_name}`?

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+  kube::build::stop_rsyncd_container
+  V=6 kube::log::status "Starting rsyncd container"
+  kube::build::run_build_command_ex \
+    "${KUBE_RSYNC_CONTAINER_NAME}" -p 127.0.0.1:${KUBE_RSYNC_PORT}:8730 -d \
```
can we make `8730` a constant somewhere?

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+    --filter='- /.make/' \
+    --filter='- /_tmp/' \
+    --filter='- /_output/' \
+    --filter='- /' \
```
is this filtering everything?

why do we have filters specified here and in rsyncd.sh and why don't they match?

jbeda replied:
This prevents it from explicitly trying to set uid/gid/perms on the root directory.  I'll add a comment.

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+  kube::build::start_rsyncd_container
+
+  local rsync_extra=""
+  if (( $KUBE_VERBOSE >= 6 )); then
```
nit: `${KUBE_VERBOSE}`

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+  fi
+
+  V=6 kube::log::status "Running rsync"
+  rsync $rsync_extra \
```
nit: `${rsync_extra}`

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+    --filter='- /vendor/' \
+    --filter='- /_temp/' \
+    --filter='+ /_output/dockerized/bin/**' \
+    --filter='+ zz_generated.*' \
```
this list smells like a source of many future bugs. :\

ixdy replied:
I don't really have a better idea (besides possibly having some process which notices changes to the volume?), but we might want to open an issue to keep track of this. having to keep a list of everything we might ever generate in random places is pretty bad.

ixdy commented on build/common.sh:
```diff
@@ -566,70 +626,134 @@ function kube::build::run_build_command() {
+  kube::build::start_rsyncd_container
+
+  local rsync_extra=""
+  if (( $KUBE_VERBOSE >= 6 )); then
```
`${KUBE_VERBOSE}`

jbeda replied:
done

ixdy commented on hack/update-generated-runtime.sh:
```diff
@@ -35,11 +35,11 @@ function prereqs() {
   KUBE_BUILD_IMAGE="${KUBE_BUILD_IMAGE_REPO}:${KUBE_BUILD_IMAGE_TAG}"
   KUBE_BUILD_CONTAINER_NAME="kube-build-${KUBE_ROOT_HASH}"
-  KUBE_BUILD_DATA_CONTAINER_NAME="kube-build-data-${KUBE_ROOT_HASH}"
+  KUBE_DATA_CONTAINER_NAME="kube-build-data-${KUBE_ROOT_HASH}"
```
should we be worried these aren't using the version?

jbeda replied:
This all goes away when I update these to use the stock build container.  I tried to separate it out in to different commits to make it a bit easier.

ixdy commented on build/release.sh:
```diff
@@ -16,7 +16,7 @@
 # Build a Kubernetes release.  This will build the binaries, create the Docker
 # images and other build artifacts.
-# For pushing these artifacts publicly on Google Cloud Storage, see the 
+# For pushing these artifacts publicly on Google Cloud Storage, see the
```
this comment should really be pointing to the kubernetes/release repo now.

ixdy reviewed (changes requested):
I still can't tell if this PR is making things more or less complicated.

jbeda committed:
Update comment on release.sh to reflect new release repo

jbeda commented:
*sigh* It looks like incremental builds are further broken by #33110.

jbeda reviewed (commented):
I think I covered everything.  I'm still figuring out the github flow and it is a bit clunky with comments on intermediate commits.

I'm pretty sure this makes things simpler.  The previous method was half broken as evidenced by the unnatural things that had to be done to copy back stuff in update-generated-*.  In addition the flow was significantly different on mac vs. linux.  Now things are very very similar so we won't have divergence there.

I'd rather just bind mount the source tree directly but that is just too slow on the mac.  I'm pretty sure that we'll need to support some level of mac development.  With that in mind, this is the best way to go.  We have a performant way to tell everyone to "just build in Docker".  It lets us punt on doing mac instructions/workarounds for the rest of the build tool chain.

As a bonus you can use a remote docker-machine from a laptop.  That is kind of nice.

We are also moving further and further down the path of extracting the release stuff out of `build/` so that we just have something that sets up a consistent docker environment.

jbeda commented:
@k8s-bot e2e test this gce e2e issue #26760

jbeda commented:
@k8s-bot gce e2e test this

jbeda commented:
@k8s-bot test this

jbeda commented:
@k8s-bot test this issue #26760

ixdy commented on build/common.sh:
```diff
@@ -378,9 +303,51 @@ function kube::build::docker_image_exists() {
+function kube::build::docker_delete_old_images() {
+  # In Docker 1.12, we can replace this with
+  #    docker images "$1" --format "{{.Tag}}"
+  for tag in $(docker images kube-build | tail -n +2 | awk '{print $2}') ; do
```
s/kube-build/$1 ?

ixdy replied:
also, should this be `${DOCKER[@]}?`

jbeda replied:
arg -- the $1 got lost when fighting docker versions.  Fixed.

ixdy commented on build/common.sh:
```diff
@@ -378,9 +303,51 @@ function kube::build::docker_image_exists() {
+function kube::build::docker_delete_old_containers() {
+  # In Docker 1.12 we can replace this line with
+  #   docker ps -a --format="{{.Names}}"
+  for container in $(docker ps -a | tail -n +2 | awk '{print $NF}') ; do
```
same re: `${DOCKER[@]}`

jbeda replied:
Done -- some of these were left over from previous code.

ixdy commented on build/common.sh:
```diff
@@ -553,19 +506,51 @@ function kube::build::ensure_data_container() {
-  kube::build::ensure_data_container
-  kube::build::prepare_output
+# Run a command in the kube-build image.  This assumes that the image has
+# already been built.  This will sync out all output data from the build.
```
comment probably needs to be updated, since this no longer syncs out? (also for `run_build_command`)

jbeda replied:
Yeah -- left from an _old_ refactor.  Updated.

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
-    cp "${RELEASE_DIR}/kubernetes-src.tar.gz" "${release_stage}/"
+# This is the port that rsync is running on *inside* the container. This may be
+# mapped to KUBE_RSYNC_PORT via docker networking.
+readonly KUBE_CONTAINER_RSYNC_PORT=8730
```
maybe move this to the top of the file with the other constants?

jbeda replied:
Done.

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
+  kube::build::stop_rsyncd_container
+  V=6 kube::log::status "Starting rsyncd container"
+  kube::build::run_build_command_ex \
+    "${KUBE_RSYNC_CONTAINER_NAME}" -p 127.0.0.1:${KUBE_RSYNC_PORT}:${KUBE_CONTAINER_RSYNC_PORT} -d \
```
does this do the right thing when `$KUBE_RSYNC_PORT` is empty?

jbeda replied:
Yeah -- that causes docker to pick an ephemeral port.  We query that with `docker port` below.

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
-    kube::release::create_tarball "${package_name}" "${release_stage}/.."
-  done
-}
+function kube::build::start_rsyncd_container() {
```
maybe comment that this sets `KUBE_RSYNC_ADDR`?

jbeda replied:
Done

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
-
-  local package_name="${RELEASE_DIR}/kubernetes-manifests.tar.gz"
-  kube::release::create_tarball "${package_name}" "${release_stage}/.."
+function kube::build::stop_rsyncd_container() {
```
maybe unset `KUBE_RSYNC_ADDR`?

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
 
-  kube::release::clean_cruft
+  V=6 kube::log::status "Running rsync"
+  rsync $rsync_extra \
```
nit: `${rsync_extra}`

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -582,462 +567,154 @@ function kube::build::run_build_command() {
-  local package_name="${RELEASE_DIR}/kubernetes.tar.gz"
-  kube::release::create_tarball "${package_name}" "${release_stage}/.."
-}
+# If the Docker server is remote, copy the results back out.
```
so this is only routinely run on macOS?

jbeda replied:
No -- now we always run this.  I'll update the comment.  The "copy out" stuff used to be only for remote/macOS docker but now it is the only way we operate.

ixdy reviewed (changes requested):
for the speedup benchmarks you posted at the beginning - is that native linux, or macOS w/ remote docker on linux?

jbeda commented on build/release.sh:
```diff
@@ -16,15 +16,17 @@
-# For pushing these artifacts publicly on Google Cloud Storage, see the 
-# associated build/push-* scripts.
+#
+# For pushing these artifacts publicly on please refer to the kubernetes/release
```
fixed

jbeda commented on build/README.md:
```diff
@@ -29,10 +27,64 @@ The following scripts are found in the `build/` directory. Note that all scripts
+
+The scripts directly under `build/` are used to build and test.  They will ensure that the `kube-build` Docker image is built (based on `build/build-image/Dockerfile`) and then execute the appropriate command in that container.  These scripts will both ensure that the right data is cached from run to run for incremental builds and will copy the results back out of the container.
+
+The `kube-build` container image is built by first creating a "context" directory in `_output/images/build-image`.  It is done there instead of at the root of the Kubernetes repo to minimize the amount of data we need to package up when building the image.
```
Adding a lot more description here.  Also describing the hash and versioning.

jbeda commented on build/README.md:
```diff
@@ -29,10 +27,64 @@ The following scripts are found in the `build/` directory. Note that all scripts
+
+The `kube-build` container image is built by first creating a "context" directory in `_output/images/build-image`.  It is done there instead of at the root of the Kubernetes repo to minimize the amount of data we need to package up when building the image.
+
+`rsync` is used transparently behind the scenes to efficiently move data in and and of the container.  This will use port an ephemeral port picked by Docker.  You can modify this by setting the `KUBE_RSYNC_PORT` env variable.
```
done

jbeda commented on build/common.sh:
```diff
@@ -55,7 +43,16 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
+# Increment/change this number if you change the build image (anything
+# under build/build-image, golang version) or change the set of volumes in the
+# data container.
+readonly KUBE_BUILD_IMAGE_VERSION=1
```
Hmmm -- that is interesting.  Right now updating `CROSS_TAG` won't cause a rebuild.  But before it was half broken as a `CROSS_TAG` update wouldn't cause data container to be blown away.

I'll update it to be a separate file and mix in CROSS_TAG.

jbeda commented:
Testing and updating changes now.

The numbers at the top of the issue are on linux.  I can't reproduce them now as something else broke incremental builds.  I'll fix that after this goes in. At this point I don't want to pile more stuff on.

jbeda commented:
Also -- sorry for the delay in responding.  I think there are still some issues with the new github review stuff.

jbeda commented:
@k8s-bot test this issue #33388

jbeda commented:
@k8s-bot test this issue #33388

ixdy commented on build/README.md:
```diff
@@ -29,10 +27,68 @@ The following scripts are found in the `build/` directory. Note that all scripts
-* `build/shell.sh`: Drop into a `bash` shell in a build container with a snapshot of the current repo code.
-* `build/release.sh`: Build everything, test it, and (optionally) upload the results to a GCS bucket.
+* `build/copy-output.sh`: This will copy the contents of `_output/dockerized/bin` from the Docker container to the local `_output/dockerized/bin`. It will also copy out specific file patterns that are generated as part of the build process. This is run automatically as part of `build/run.sh`.
+* `build/make-clean.sh`: Clean out the contents of `_output`, remove any locally built container images and remove the data container.
```
should this be removing the generated stuff that's copied out by copy-output?

ixdy commented on build/README.md:
```diff
@@ -29,10 +27,68 @@ The following scripts are found in the `build/` directory. Note that all scripts
+
+## Basic Flow
+
+The scripts directly under `build/` are used to build and test.  They will ensure that the `kube-build` Docker image is built (based on `build/build-image/Dockerfile`) and then execute the appropriate command in that container.  These scripts will both ensure that the right data is cached from run to run for incremental builds and will copy the results back out of the container
```
nittiest of nits: missing period at end of line.

jbeda replied:
done

ixdy commented on build/build-image/rsyncd.sh:
```diff
@@ -0,0 +1,86 @@
+VOLUME=${HOME}
+
+# Find the default network device
+DEFAULT_DEV=$(ip route | awk '/^default via \S+ dev \S+\s*$/ { print $5 }')
```
it looks like this isn't used anymore?

jbeda replied:
removed -- incremented BUILD_IMAGE_VERSION to account for it.

ixdy commented on build/build-image/rsyncd.sh:
```diff
@@ -0,0 +1,86 @@
+if [[ -f "${PIDFILE}" ]]; then
+  PID=$(cat "${PIDFILE}")
+  echo "Cleaning up old PID file: ${PIDFILE}"
+  killall $PID &> /dev/null || true
```
`kill` instead of `killall`?

jbeda replied:
done

ixdy commented on build/common.sh:
```diff
@@ -55,7 +43,17 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
+# Jenkins.
+#
+# Increment/change this number if you change the build image (anything
+# under build/build-image, golang version) or change the set of volumes in the
```
probably don't need to change this for golang version anymore, right?

jbeda replied:
Yeah, probably right.

jbeda commented:
PTAL -- hopefully we are in the home stretch here.  Thanks.

jbeda commented:
@ixdy -- ping?!?  I'd really love to get this in.  I'm around to be reactive and fix any issues that might crop up.

ixdy commented on build/common.sh:
```diff
@@ -55,7 +43,16 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
 readonly KUBE_BUILD_IMAGE_REPO=kube-build
 readonly KUBE_BUILD_IMAGE_CROSS_TAG="$(cat ${KUBE_ROOT}/build/build-image/cross/VERSION)"
-# KUBE_BUILD_DATA_CONTAINER_NAME=kube-build-data-<hash>"
+# KUBE_DATA_CONTAINER_NAME=kube-build-data-<hash>"
```
maybe just remove this?

jbeda replied:
Yeah - this is left over.  I was documenting the constants up here but some of them aren't really constant and are set in `kube::build::verify_prereqs` now.  There is enough stuff there that it should be cool deleting this comment.

ixdy commented on build/common.sh:
```diff
@@ -55,7 +43,16 @@ readonly KUBE_BUILD_PPC64LE="${KUBE_BUILD_PPC64LE:-n}"
+# KUBE_DATA_CONTAINER_NAME=kube-build-data-<hash>"
+
+# This version number is used to cause everyone to rebuild their data containers
+# and build image.  This is especially useful for automated build systems like
```
note that on Jenkins we start with a `make clean`, so this is less important there.

jbeda replied:
Hmm -- Tim had some issues with this at one point where he was mucking this stuff out.  But we should still do this as it'll trip up others.

ixdy commented on build/common.sh:
```diff
@@ -409,70 +382,57 @@ function kube::build::destroy_container() {
-  mkdir -p "${LOCAL_OUTPUT_BUILD_CONTEXT}"
-  "${TAR}" czf "${LOCAL_OUTPUT_BUILD_CONTEXT}/kube-source.tar.gz" $(kube::build::source_targets)
+    kube::version::get_version_vars
+    kube::version::save_version_vars "${LOCAL_OUTPUT_BUILD_CONTEXT}/kube-version-defs"
```
am I reading this correctly? we won't update the version info embedded in binaries unless we `make clean` first?

ixdy replied:
should this be saved in the data container instead?

jbeda replied:
Dammit!  This is a good catch.  I'll move this to a file under `KUBE_ROOT` so it gets copied with the rsync.

ixdy reviewed (commented):
Basically LG, or at least I'm happy to merge and then iterate with further improvements. Just a few comments.

ixdy commented on build/lib/release.sh:
```diff
@@ -0,0 +1,450 @@
+}
+
+###############################################################################
+# Most of the ::release:: namespace functions have been moved to
```
it might be worth moving this comment to the top of the file.

a comment describing what this file is for (packaging all release artifacts) might not be a bad idea, either.

jbeda replied:
Done -- also made a change to remove a bunch of unused variables in this file.  That is a new commit in this PR.

jbeda committed:
Remove unused bash variables

jbeda commented:
Okay! Please take another look.  Hopefully we are there.

ivan4th commented:
I have slightly unexpected possible use for remote docker mode: don't rsync back binaries after build, use them to start [DIND cluster](https://github.com/sttts/kubernetes-dind-cluster) on remote docker instead. E.g. I'm on a not very fast 3G connection and want to use GCE for dev. How hard it's to do this? (at least opt out of rsyncing binaries back)

jbeda commented:
@ivan4th -- interesting case!  Right now we always "copy back" to the host machine but we can explore making that optional after this gets in.

ivan4th commented:
@jbeda actually it looks like `KUBE_RUN_COPY_OUTPUT=n build/run.sh` may do the trick.

ixdy commented on build/lib/release.sh:
```diff
@@ -0,0 +1,442 @@
+# limitations under the License.
+
+# This file creates release artifacts (tar files, container images) that are
+# ready to distribute to install or distribute to end users.
```
nit: "ready to distribute to install or distribute" seems awkward?

ixdy reviewed (approved):
woo! I'm fine with getting this in and then iterating with any additional fixes if necessary. two minor comments, neither of which is blocking.

duglin commented:
woo hoo!
